src/ae_kqueue.c

```diff
@@ -133,22 +140,26 @@ static int aeApiPoll(aeEventLoop *eventLoop, struct timeval *tvp) {
          * the same fd events later. */
         for (j = 0; j < retval; j++) {
             struct kevent *e = state->events+j;
+            int fd = e->ident;
+            int mask = 0; 
 
-            if (e->filter == EVFILT_READ) state->eventsMask[e->ident] |= AE_READABLE;
-            else if (e->filter == EVFILT_WRITE) state->eventsMask[e->ident] |= AE_WRITABLE;
+            if (e->filter == EVFILT_READ) mask = AE_READABLE;
+            else if (e->filter == EVFILT_WRITE) mask = AE_WRITABLE;
+            state->eventsMask[fd/4] |= EVENT_MASK_ENCODE(fd, mask);
         }
 
         /* Re-traversal to merge read and write events, and set the fd's mask to
          * 0 so that events are not added again when the fd is encountered again. */
         numevents = 0;
         for (j = 0; j < retval; j++) {
             struct kevent *e = state->events+j;
-            int mask = state->eventsMask[e->ident];
+            int fd = e->ident;
+            int mask = EVENT_MASK_DECODE(fd, state->eventsMask[fd/4]);
 
             if (mask) {
-                eventLoop->fired[numevents].fd = e->ident;
+                eventLoop->fired[numevents].fd = fd;
                 eventLoop->fired[numevents].mask = mask;
-                state->eventsMask[e->ident] = 0;
+                state->eventsMask[fd/4] &= ~EVENT_MASK_ENCODE(fd, AE_READABLE|AE_WRITABLE);
```

Comment:
i now realize that we rely on the previous iteration to clean after itself, which means that we also need a memset(0) after malloc and realloc

Reply:
Done, Thanks for the reminder.